### PR TITLE
avoid creating `PyRef` inside `__traverse__` handler

### PR DESCRIPTION
**Pull request by davidhewitt**
See https://github.com/PyO3/pyo3/issues/4265#issuecomment-2305890021

We cannot use `PyRef<T>` to guard access to the value inside a `__traverse__` handler. Instead I use the `PyClassObject<T>` and manually perform the borrow checking.

### CODE REVIEW

File: src/pyclass/gc.rs [hunk 1]
```diff
@@ -1,20 +1,29 @@
...
+    os::raw::{c_int, c_void},
+};
 
-use crate::{ffi, AsPyPointer, Python};
+use crate::{ffi, AsPyPointer};
 
 /// Error returned by a `__traverse__` visitor implementation.
 #[repr(transparent)]
-pub struct PyTraverseError(pub(crate) c_int);
+pub struct PyTraverseError(get_nonzero_c_int::TYPE);
```
Comment (davidhewitt) [inline src/pyclass/gc.rs:11]: A bit of a drive-by optimization, to add a niche for the 0.

File: src/pyclass/gc.rs [hunk 2]
```diff
@@ -1,20 +1,29 @@
...
-pub struct PyVisit<'p> {
+pub struct PyVisit<'a> {
     pub(crate) visit: ffi::visitproc,
     pub(crate) arg: *mut c_void,
-    /// VisitProc contains a Python instance to ensure that
-    /// 1) it is cannot be moved out of the traverse() call
-    /// 2) it cannot be sent to other threads
-    pub(crate) _py: Python<'p>,
+    /// Prevents the `PyVisit` from outliving the `__traverse__` call.
+    pub(crate) _guard: PhantomData<&'a ()>,
```
Comment (davidhewitt) [inline src/pyclass/gc.rs:26]: I completely removed the `Python::assume_gil_acquired` call from `_call_traverse`, so I needed to place a different reference back in for the lifetime. This is sufficient because the `*mut c_void` already prevents `Send`/`Sync`.

**Comment by davidhewitt** [comment]
cc @ngoldbaum

**Comment by mejrs** [comment]
We could also make the change suggested by the comment here....
```rust
#[repr(transparent)]
pub struct PyRef<'p, T: PyClass> {
    // TODO: once the GIL Ref API is removed, consider adding a lifetime parameter to `PyRef` to
    // store `Borrowed` here instead, avoiding reference counting overhead.
    inner: Bound<'p, T>,
}
```
 as that would also avoid changing reference counts.

**Comment by Icxolu** [comment]
I believe that would need #4390. Also in that thread https://github.com/PyO3/pyo3/pull/4390#discussion_r1702907653 there was the suggestion to maybe delay that change to 0.24

**Comment by davidhewitt** [comment]
Yes agreed, I considered changing `PyRef` but would prefer to delay given there's enough breaking changes coming in 0.23. I also want to be careful with changing `PyRef` right now because I think it might need other changes for the freethreaded build. 

I also think I might backport this fix to 0.22, which would be another reason not to have the breaking change here.

**Comment by ngoldbaum** [comment]
I did some local testing on this PR. 

After rebasing on `main` to fix some noisy tests that were fixed by me last week, this does seem to fix the `"accessing the GIL while a __traverse__  implementation is running"` panics.

**Comment by davidhewitt** [comment]
Great, will merge this. I am considering landing this as part of a stack of fixes in a 0.22.3 patch.
